state-chain/pallets/cf-lending-pools/src/general_lending.rs

```diff
@@ -0,0 +1,1635 @@
+	/// amount.
+	fn repay_funds(&mut self, amount: AssetAmount) {
+		GeneralLendingPools::<T>::mutate(self.asset, |maybe_pool| {
+			if let Some(pool) = maybe_pool.as_mut() {
+				pool.receive_repayment(amount);
+			} else {
+				log_or_panic!("CHP Pool must exist for asset {}", self.asset);
+			}
+		});
+
+		self.owed_principal.saturating_reduce(amount);
+	}
+
+	/// Pays fee to the pool in pool's asset. Does NOT reduce the owed principal.
+	fn pay_fee(&mut self, amount: AssetAmount) {
+		GeneralLendingPools::<T>::mutate(self.asset, |maybe_pool| {
+			if let Some(pool) = maybe_pool.as_mut() {
+				pool.receive_fees(amount);
+			} else {
+				log_or_panic!("CHP Pool must exist for asset {}", self.asset);
+			}
+		});
+	}
+}
+
+fn get_price<T: Config>(asset: Asset) -> Result<Price, Error<T>> {
+	Ok(T::PriceApi::get_price(asset).ok_or(Error::<T>::OraclePriceUnavailable)?.price)
+}
+
+/// Uses oracle prices to calculate the amount of `asset_2` that's equivalent in USD value to
+/// `amount` of `asset_1`
+fn equivalent_amount<T: Config>(
+	asset_1: Asset,
+	asset_2: Asset,
+	amount: AssetAmount,
+) -> Result<AssetAmount, Error<T>> {
+	let asset_1_price = get_price::<T>(asset_1)?;
+	let asset_2_price = get_price::<T>(asset_2)?;
+
+	// how much of asset 2 you get per asset 1
+	let price = relative_price(asset_1_price, asset_2_price);
+
+	Ok(cf_amm_math::output_amount_ceil(amount.into(), price).unique_saturated_into())
+}
+
+/// Uses oracle prices to calculate the USD value of the given asset amount
+fn usd_value_of<T: Config>(asset: Asset, amount: AssetAmount) -> Result<AssetAmount, Error<T>> {
+	let price_in_usd = get_price::<T>(asset)?;
+	Ok(cf_amm_math::output_amount_ceil(amount.into(), price_in_usd).unique_saturated_into())
+}
+
+/// Uses oracle prices to calculate the amount of `asset` that's equivalent in USD value to
+/// `amount` of USD
+fn amount_from_usd_value<T: Config>(
+	asset: Asset,
+	usd_value: AssetAmount,
+) -> Result<AssetAmount, Error<T>> {
+	// The "price" of USD in terms of the asset:
+	let price = invert_price(get_price::<T>(asset)?);
+	Ok(cf_amm_math::output_amount_ceil(usd_value.into(), price).unique_saturated_into())
+}
+
+/// A wrapper around `init_swap_request` that uses parameter suitable for a lending pool swap
+fn initiate_swap<T: Config>(
+	from_asset: Asset,
+	amount: AssetAmount,
+	to_asset: Asset,
+	swap_type: LendingSwapType<T::AccountId>,
+	max_oracle_price_slippage: BasisPoints,
+) -> SwapRequestId {
+	T::SwapRequestHandler::init_swap_request(
+		from_asset,
+		amount,
+		to_asset,
+		SwapRequestType::Regular {
+			output_action: SwapOutputAction::CreditLendingPool { swap_type },
+		},
+		Default::default(), // broker fees
+		Some(PriceLimitsAndExpiry {
+			expiry_behaviour: ExpiryBehaviour::NoExpiry,
+			min_price: Default::default(),
+			max_oracle_price_slippage: Some(max_oracle_price_slippage),
+		}),
+		None, // no dca parameters
```

Comment:
If we never use DCA, does this mean liquidations are always all-or-nothing? I would have expected we might want to use DCA for liquidations but not for fee swaps, for example.

Reply:
Right, I think I was delaying enabling DCA until we decide what the parameters should be... I guess we can define a fixed chunk size controlled by governance? For simplicity could be the same value for each asset.